src/corefx/System.Globalization.Native/collation.cpp

```diff
@@ -126,6 +126,84 @@ extern "C" int32_t LastIndexOf(
 }
 
 /*
+Static Function:
+AreEqualOrdinalIgnoreCase
+*/
+static bool AreEqualOrdinalIgnoreCase(UChar one, UChar two)
+{
+	// Return whether the two characters are identical or would be identical if they were upper-cased.
+
+	if (one == two)
+	{
+		return true;
+	}
+
+	if (one == 0x0131 || two == 0x0131)
+	{
+		// On Windows with InvariantCulture, the LATIN SMALL LETTER DOTLESS I (U+0131)
+		// capitalizes to itself, whereas with ICU it capitalizes to LATIN CAPITAL LETTER I (U+0049).
+		// We special case it to match the Windows invariant behavior.
+		return false;
+	}
+
+	return u_toupper(one) == u_toupper(two);
+}
+
+/*
+Function:
+IndexOfOrdinalIgnoreCase
+*/
+extern "C" int32_t
+IndexOfOrdinalIgnoreCase(const UChar* lpTarget, int32_t cwTargetLength, const UChar* lpSource, int32_t cwSourceLength)
+{
+	int32_t endIndex = cwSourceLength - cwTargetLength;
+	assert(endIndex >= 0);
+
+	for (int32_t i = 0; i <= endIndex; i++)
+	{
+		int32_t targetIdx = 0;
+		for (int32_t srcIdx = i; targetIdx < cwTargetLength; srcIdx++, targetIdx++) {
+			if (!AreEqualOrdinalIgnoreCase(lpSource[srcIdx], lpTarget[targetIdx])) {
```

Comment:
The deseret letters are outside the BMP and have casing data.  You could use:

'DESERET CAPITAL LETTER LONG I' (U+10400) and 'DESERET SMALL LETTER LONG I' (U+10428) as a case pair.

I just looked at the windows behavior and it confuses me a bit.

I would expect that the following would return 0:

`"\U00010428".IndexOf("\U00010400", StringComparison.OrdinalIgnoreCase);`

But, on Windows 10 (on Full .NET) this returns -1.  The "equivalent" code of:

`"\U00010428".ToUpperInvariant().IndexOf("\U00010400".ToUpperInvariant(), StringComparison.Ordinal);`

Returns "0" as expected.

Per [MSDN](https://msdn.microsoft.com/en-us/library/system.stringcomparer.ordinalignorecase(v=vs.110).aspx):

> OrdinalIgnoreCase property treats the characters in the strings to compare as if they were converted to uppercase using the conventions of the invariant culture, and then performs a simple byte comparison that is independent of language.

You can generate similar cases for string equality where OridinalIgnoreCase does not behave the same way vs Ordinal with a ToUpperInvariant beforehand.

It's possible that the windows routines don't do the right thing with characters outside the BMP.  I'm unsure if we should be copying that behavior, however.  It would be interesting to know what @tarekgh thinks.